File: packages/sim/src/terminals/generators/lan.ts
import type { MatchQuestion } from "../types";
import { base, param, pool, randomMac, range, type GenContext } from "./common";

interface LanPool {
  address_format: {
    prompt: string;
    types: Record<string, string>;
    domains: string[];
    ports: number[];
    explanation: string;
    hint: string;
  };
  switch_table: {
    prompt: string;
    devices: string[];
    inventoryTitle: string;
    arpTitle: string;
    camTitle: string;
    headers: { inventory: string[]; arp: string[]; cam: string[] };
    portLabel: string;
    explanation: string;
    hint: string;
  };
}

/** Associação valor ↔ tipo de identificador (IPv4, MAC, porta, domínio, IPv6). */
export function addressFormatMatch(ctx: GenContext): MatchQuestion {
  const P = pool<LanPool>(ctx, "lan").address_format;
  const rng = ctx.rng;
  const types = param<string[]>(ctx, "types", ["ipv4", "mac", "port", "domain"]);
  const values = types.map((t) => {
    switch (t) {
      case "ipv4": return `192.168.${rng.int(0, 20)}.${rng.int(2, 254)}`;
      case "mac": return randomMac(rng);
      case "port": return String(rng.pick(P.ports));
      case "domain": return rng.pick(P.domains);
      case "ipv6": return `fe80::${rng.int(0, 65535).toString(16)}:${rng.int(0, 65535).toString(16)}`;
      default: throw new Error(`unknown address type ${t}`);
    }
  });
  const idx = types.map((_, i) => i);
  const leftOrder = rng.shuffle(idx);
  const rightOrder = rng.shuffle(idx);
  return {
    ...base(ctx, "address_format_match"),
    kind: "match",
    prompt: P.prompt,
    left: leftOrder.map((i) => values[i]!),
    right: rightOrder.map((i) => P.types[types[i]!]!),
    answer: leftOrder.map((i) => rightOrder.indexOf(i)),
    explanation: P.explanation,
    hint: P.hint,
  };
}

/** Localizar a porta do switch de cada dispositivo cruzando inventário (nome→IP), ARP (IP→MAC) e tabela MAC (porta→MAC). */
export function switchTableMatch(ctx: GenContext): MatchQuestion {
  const P = pool<LanPool>(ctx, "lan").switch_table;
  const rng = ctx.rng;
  const count = param(ctx, "count", 4);
  const octet = rng.int(1, 30);
  const names = rng.sample(P.devices, count);
  const hosts = rng.sample(range(2, 200), count);
  const macs = names.map(() => randomMac(rng));
  const ports = rng.sample(range(1, 8), count);
  const ips = hosts.map((h) => `192.168.${octet}.${h}`);
  const idx = names.map((_, i) => i);
  const arpOrder = rng.shuffle(idx);
  const camOrder = rng.shuffle(idx);
  const sortedPorts = ports.slice().sort((a, b) => a - b);
  const label = (n: number): string => P.portLabel.replace("{n}", String(n));
  return {
    ...base(ctx, "switch_table_match"),
    kind: "match",
    prompt: P.prompt,
    context: [
      { type: "table", title: P.inventoryTitle, headers: P.headers.inventory, rows: idx.map((i) => [names[i]!, ips[i]!]) },
      { type: "table", title: P.arpTitle, headers: P.headers.arp, rows: arpOrder.map((i) => [ips[i]!, macs[i]!]) },
      { type: "table", title: P.camTitle, headers: P.headers.cam, rows: camOrder.map((i) => [label(ports[i]!), macs[i]!]) },
    ],
    left: idx.map((i) => names[i]!),
    right: sortedPorts.map(label),
    answer: idx.map((i) => sortedPorts.indexOf(ports[i]!)),
    explanation: P.explanation,
    hint: P.hint,
  };
}

File: apps/api/app/terminals/generators/lan.py
"""Geradores da rede local — espelho de generators/lan.ts."""

from __future__ import annotations

from typing import Any

from .common import GenContext, base, int_range, param, pool, random_mac


def address_format_match(ctx: GenContext) -> dict[str, Any]:
    p = pool(ctx, "lan")["address_format"]
    rng = ctx.rng
    types: list[str] = param(ctx, "types", ["ipv4", "mac", "port", "domain"])
    values: list[str] = []
    for t in types:
        if t == "ipv4":
            values.append(f"192.168.{rng.int(0, 20)}.{rng.int(2, 254)}")
        elif t == "mac":
            values.append(random_mac(rng))
        elif t == "port":
            values.append(str(rng.pick(p["ports"])))
        elif t == "domain":
            values.append(rng.pick(p["domains"]))
        elif t == "ipv6":
            values.append(f"fe80::{format(rng.int(0, 65535), 'x')}:{format(rng.int(0, 65535), 'x')}")
        else:
            raise ValueError(f"unknown address type {t}")
    idx = list(range(len(types)))
    left_order = rng.shuffle(idx)
    right_order = rng.shuffle(idx)
    return {
        **base(ctx, "address_format_match"),
        "kind": "match",
        "prompt": p["prompt"],
        "left": [values[i] for i in left_order],
        "right": [p["types"][types[i]] for i in right_order],
        "answer": [right_order.index(i) for i in left_order],
        "explanation": p["explanation"],
        "hint": p["hint"],
    }


def switch_table_match(ctx: GenContext) -> dict[str, Any]:
    p = pool(ctx, "lan")["switch_table"]
    rng = ctx.rng
    count = param(ctx, "count", 4)
    octet = rng.int(1, 30)
    names = rng.sample(p["devices"], count)
    hosts = rng.sample(int_range(2, 200), count)
    macs = [random_mac(rng) for _ in names]
    ports = rng.sample(int_range(1, 8), count)
    ips = [f"192.168.{octet}.{h}" for h in hosts]
    idx = list(range(len(names)))
    arp_order = rng.shuffle(idx)
    cam_order = rng.shuffle(idx)
    sorted_ports = sorted(ports)
    headers = p["headers"]

    def label(n: int) -> str:
        return p["portLabel"].replace("{n}", str(n), 1)

    return {
        **base(ctx, "switch_table_match"),
        "kind": "match",
        "prompt": p["prompt"],
        "context": [
            {"type": "table", "title": p["inventoryTitle"], "headers": headers["inventory"], "rows": [[names[i], ips[i]] for i in idx]},
            {"type": "table", "title": p["arpTitle"], "headers": headers["arp"], "rows": [[ips[i], macs[i]] for i in arp_order]},
            {"type": "table", "title": p["camTitle"], "headers": headers["cam"], "rows": [[label(ports[i]), macs[i]] for i in cam_order]},
        ],
        "left": [names[i] for i in idx],
        "right": [label(n) for n in sorted_ports],
        "answer": [sorted_ports.index(ports[i]) for i in idx],
        "explanation": p["explanation"],
        "hint": p["hint"],
    }
